extension/extensionlimiter/extensionlimiter.go

```diff
@@ -0,0 +1,71 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package extensionlimiter // import "go.opentelemetry.io/collector/extension/extensionlimiter"
+
+import (
+	"context"
+)
+
+// Predefined weight keys for common rate limits
+const (
+	WeightKeyNetworkBytes   = "network_bytes"
+	WeightKeyRequestItems   = "request_items"
+	WeightKeyRequestCount   = "request_count"
+	WeightKeyResidentMemory = "resident_memory"
+)
+
+// Weight represents a specific dimension to be limited with its associated value.
+type Weight struct {
+	// Key identifies the type of weight being limited
+	Key string
```

Comment:
Should we use an enum here instead?

Reply:
Done. (TODO: I believe I need an Unmarshal function)

Reply:
Alternatively, maybe we could get rid of the key, and make it part of the caller's configuration to specify different limiters? https://github.com/open-telemetry/opentelemetry-collector/pull/12700#discussion_r2011318894

Reply:
I have changed the API. I believe it is best to use an open set for these values, with the four existing keys defined for now.

In the current form, the limiter extension API is a Provider, and this allows components to perform a nil-check at startup, making it possible to bypass overhead associated with calculating weights when there is no limiter subscribed. This makes it easy to fail at startup if a limiter is configured with a non-existing limit name, and this makes it possible to use arbitrary limiter names later, if we decide.